src/frontend/src/binder/query.rs

```diff
@@ -30,6 +31,7 @@ pub struct BoundQuery {
     pub order: Vec<FieldOrder>,
     pub limit: Option<usize>,
     pub offset: Option<usize>,
+    pub order_exprs: Vec<ExprImpl>,
```

Comment:
* `extra_order_exprs`
* Maybe we can put this on `BoundSelect` rather than `BoundQuery`, because this feature only works when `body` matches `BoundSetExpr::Select`. They cannot be used with `VALUES`, `UNION`, or even `SELECT DISTINCT`.

Side note: as of now we push and pop context in `bind_query`. But in `select a1 from a union select b1 from b`, the context should actually be different for each `select`. We do not need to fix in this PR, but this is why we cannot bind `order by a1` here - it is only available in the context of the `select`.

Reply:
I orginally wanted to put it in `BoundSelect`, but ORDER BY clause is bound after `BoundSelect` was generated and because BoundSetExpr is an enum, it is difficult to put `extra_order_exprs` into `BoundSelect`.

Reply:
It is not difficult as matching enum also helps reject `union` here.

However, after rethinking, I agree that putting `extra_order_exprs` at the same level of `order` and `out_fields` is better. It can be confusing that each `BoundSelect` has `extra_order_exprs`, which is only meaningful when it has no siblings.

As for rejecting `union` as well as using the correct context when binding `extra_order_exprs`, the same idea can be extended from `values`: (1) Without `union`, the `query` and its body `select` share the same context (current behavior); (2) With `union`, each descendant `select` will use their own context, leaving the context of `query` empty, and `extra_order_exprs` would return bind error as expected.